File: src/3.c
// The prime factors of 13195 are 5, 7, 13, and 29.
// 
// What is the largest prime factor of the number 600851475143?
# include <stdio.h>

int isPrime(long number){
    for(long i = 2; i < number; i++){
        if(number % i == 0){
            return 0;
        }
    }
    return 1;
}

void main() {
    const long C = 600851475143;
    long c = C;
    for(long i = 2; i < C; i++){
        if(c % i == 0 && isPrime(i)) {
            c /= i;
            if(c == 1){
                printf("%ld\n", i);//6857
                break;
            }
        }
    }
}

File: src/0-9/3.ml
open Stdio

(* Finds the largest prime factor of a given number *)
let largest_prime_factor number =
  let rec factor n i =
    if n = 1 then i 
    else if n mod i = 0 then factor (n / i) i
    else factor n (i + 1)
  in
  factor number 2

let () =
  let number_to_factor = 600851475143 in
  printf "The largest prime factor of %d is %d\n"
    number_to_factor (largest_prime_factor number_to_factor)
